### DISCUSSION TITLE
StatAnalysis OBS_SID -aggregate Issue

### DISCUSSION THREAD

**Discussion by lindsayrblank** [dtcenter/METplus#1192] [Statistical Computation] [answered]
Hi wonderful MET folks,

I'm running into some unexpected behavior from StatAnalysis based on what I've seen before. I am trying to aggregate multiple partial sums together and would like to aggregate across forecast lead, forecast variable, forecast init, and obs_sid. For example, all 6 hour lead times for surface temperature initialized at 00 UTC at KDEN. When I run the following command:

singularity exec path/to/image/met_bugfix_1875.sif stat_analysis -lookin <directory> -job aggregate -line_type SL1L2 -out_stat <out_file_name> -fcst_init_hour 00 -by OBS_SID,FCST_LEAD,FCST_VAR -set_hdr VX_MASK OBS_SID -rank_corr_flag FALSE -n_boot_rep 0 -out_bin_size 1 -v 3

I get lines like this: '-9999:010000:VariableName'

All of the obs_sids have been combined into one, whereas the other two 'FCST_LEAD' and 'FCST_VAR' are aggregated as intended. 

I will get debug output like " For case "-9999:050000:VariableName", found 4288 unique VX_MASK values:" and then list all of the station IDs. 

I'm running StatAnalysis on a HPC via Singularity starting from the DTC Docker image met_bugfix_1875.sif. I am using version 10.0.0.

Thank you in advance for your help!

**Comment by JohnHalleyGotway** [comment, marked as answer]
@lindsayrblank I see you have a question about the STAT-Analysis tool. Looking carefully at the job you're running, I see the issue. It's not a bug. You're just requesting data that doesn't actually exist.

Your job (listed below) aggregates (-job aggregate) SL1L2 partial sum lines (-line_type SL1L2). The -by option requests that the aggregation be performed for each unique combination of OBS_SID/FCST_LEAD/FCST_VAR columns found in the input.
```
-job aggregate -line_type SL1L2 -out_stat <out_file_name> -fcst_init_hour 00 -by OBS_SID,FCST_LEAD,FCST_VAR -set_hdr VX_MASK OBS_SID
```
For aggregation jobs, the input line type = output line type. And the -by columns are read from those input lines. However, the input SL1L2 line type does NOT include the OBS_SID column. Since that column doesn't exist, STAT-Analysis is just returning a bad data string value (-9999).

I wonder if perhaps you actually intended to run an aggregate_stat job type to aggregate matched pair (MPR) input lines into SL1L2 partial sum output lines? Perhaps something like this:
```
-job aggregate_stat -line_type MPR -out_line_type SL1L2 -out_stat <out_file_name> -fcst_init_hour 00 -by OBS_SID,FCST_LEAD,FCST_VAR -set_hdr VX_MASK OBS_SID -rank_corr_flag FALSE -n_boot_rep 0 -out_bin_size 1 -v 3
```
This job would work better since the OBS_SID column actually does exist in the input [MPR line type](https://met.readthedocs.io/en/latest/Users_Guide/point-stat.html?highlight=MPR#id24).

For completeness sake, here's [documentation](https://met.readthedocs.io/en/latest/Users_Guide/stat-analysis.html#aggregate-stat-lines-and-produce-aggregated-statistics) about the aggregate_stat job type.

**Reply by lindsayrblank** [reply]
Hi John, I understand now. Thank you for explaining that! I used -aggregate_stat to create SL1L2 output from MPR first, then was trying to aggregate multiple SL1L2 files together. Since I set VX_MASK to be OBS_SID in the SL1L2 output, it sounds like should set '-by VX_MASK' instead of '-by OBS_SID' when aggregating my SL1L2 output together in order to get the output I desire. 

I just tried it and it provides the output I desire. Thank you for catching my mistake!

**Reply by JohnHalleyGotway** [reply]
Great Lindsay, thanks for confirming that worked. I'm going to mark this Discussion as "Answered" and also "Lock the Conversation" to prevent future posts. If more issues or questions arise, please just start a new Discussion topic.
